hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/kafka/SqlAvroTest.java

```diff
@@ -401,6 +398,134 @@ public void test_nonInclusiveUnion() {
                 "Not in union [\"null\",\"boolean\",\"int\"]: " + Long.MAX_VALUE + " (Long) (field=info)");
     }
 
+    @Test
+    public void test_fieldResolution_unsupportedSchemaTypes() {
+        assumeFalse(useSchemaRegistry);
+
+        List<Schema> unsupportedSchemaTypes = List.of(
+                Schema.create(Schema.Type.BYTES),
+                SchemaBuilder.array().items(Schema.create(Schema.Type.INT)),
+                SchemaBuilder.map().values(Schema.create(Schema.Type.INT)),
+                SchemaBuilder.enumeration("enum").symbols("symbol"),
+                SchemaBuilder.fixed("fixed").size(0)
+        );
+        for (Schema schema : unsupportedSchemaTypes) {
+            Schema valueSchema = SchemaBuilder.record("jet.sql").fields()
+                    .name("field").type(schema).noDefault()
+                    .endRecord();
+            assertThatThrownBy(() -> kafkaMapping("kafka", ID_SCHEMA, valueSchema).create())
+                    .hasMessage("Unsupported schema type: " + schema.getType());
+        }
+    }
+
+    @Test
+    public void test_fieldResolution() {
+        assumeFalse(useSchemaRegistry);
+
+        // Test field resolution for top-level fields
+        Schema childSchema = SchemaBuilder.record("child").fields()
+                .requiredBoolean("requiredBoolean")
+                .nullableBoolean("nullableBoolean", true)
+                .optionalBoolean("optionalBoolean")
+                .requiredInt("requiredInt")
+                .nullableInt("nullableInt", 1)
+                .optionalInt("optionalInt")
+                .requiredLong("requiredLong")
+                .nullableLong("nullableLong", 1L)
+                .optionalLong("optionalLong")
+                .requiredFloat("requiredFloat")
+                .nullableFloat("nullableFloat", 1F)
+                .optionalFloat("optionalFloat")
+                .requiredDouble("requiredDouble")
+                .nullableDouble("nullableDouble", 1D)
+                .optionalDouble("optionalDouble")
+                .requiredString("requiredString")
+                .nullableString("nullableString", "A")
+                .optionalString("optionalString")
+                .name("requiredUnion").type().unionOf().intType().and().stringType().endUnion().noDefault()
+                .name("nullableUnion").type().unionOf().intType().and().stringType().and().nullType().endUnion()
+                        .intDefault(1)
+                .name("optionalUnion").type().unionOf().nullType().and().intType().and().stringType().endUnion()
+                        .nullDefault()
+                .endRecord();
+
+        String name = createRandomTopic();
+        kafkaMapping(name, ID_SCHEMA, childSchema).create();
```

Comment:
general comment for consideration, not necessarily in this PR: `ID_SCHEMA` and `NAME_SCHEMA` use `jet.sql` record name. Cases where avro schema is given should work with any record name. Is there a risk that we rely on special handling of `jet.sql` record name (if not specified, it is default)?

Reply:
When the key/value Avro schema is provided, its name is irrelevant for `AvroUpsertTarget`, similarly to field default values, which are stored in `Schema` and handled by `GenericRecordBuilder`.

When resolving schema from columns, a record name is needed, which defaults to "jet.sql" if unspecified. I think this doesn't make sense since there can be at most one schema with this name for the reader to able to differentiate schemas. I want to make `(key|value)AvroRecordName` mandatory for cases that trigger schema resolution. Also, I want to introduce `avroRecordName` for UDTs to override nested record names.

I also realized that `KvMetadataAvroResolver` assumes the schema exists in field resolution, and fields exist in schema resolution. They should be fixed and tested.

Reply:
> at most one schema with this name for the reader to able to differentiate schemas

if users know which `jet.sql` schema use for which topic, this can be managed. I believe that is also the way how sql->kafka avro->sql would work without explicit schemas and without schema registry. So limited, but doable.

> I want to make (key|value)AvroRecordName mandatory for cases that trigger schema resolution

It would remove current behavior of using default `jet.sql` so it would be a breaking change.  IMO it is not the best feature but currently it seems not to be harmful.

Reply:
HZ-3705